File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.java
class Solution {
    public int subarraySum(int[] nums, int k) {
        Map<Integer, Integer> sumFreq = new HashMap<>();
        sumFreq.put(0, 1);
        int count = 0;
        int sum = 0;
        int n = nums.length;
        for(int i = 0; i < n; i++){
            sum = sum + nums[i];
            
            int compliment = sum - k;
            
            if(sumFreq.containsKey(compliment)){
                count = count + sumFreq.get(compliment);
            }
            
            sumFreq.put(sum, sumFreq.getOrDefault(sum, 0)+1);
        }
        return count;
    }
}

File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.cpp
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        unordered_map<int, int> sumFreq;
        sumFreq[0] = 1;
        int sum = 0;
        int count = 0;
        int n = nums.size();
        for(int i=0; i<n; i++){
            sum = sum + nums[i];
            
            int compliment = sum - k;
            
            if(sumFreq.find(compliment) != sumFreq.end()){
                count = count + sumFreq[compliment];
            }
            
            sumFreq[sum]++;
        }
        return count;
    }
};